lib/util/glob-util.js

```diff
@@ -95,16 +99,22 @@ const dotfilesPattern = /(?:(?:^\.)|(?:[/\\]\.))[^/\\.].*/;
  * Build a list of absolute filesnames on which ESLint will act.
  * Ignored files are excluded from the results, as are duplicates.
  *
- * @param   {string[]} globPatterns                    Glob patterns.
- * @param   {Object}   [providedOptions]               An options object.
- * @param   {string}   [providedOptions.cwd]           CWD (considered for relative filenames)
- * @param   {boolean}  [providedOptions.ignore]        False disables use of .eslintignore.
- * @param   {string}   [providedOptions.ignorePath]    The ignore file to use instead of .eslintignore.
- * @param   {string}   [providedOptions.ignorePattern] A pattern of files to ignore.
+ * @param   {string[]} globPatterns                     Glob patterns.
+ * @param   {Object}   [providedOptions]                An options object.
+ * @param   {string}   [providedOptions.cwd]            CWD (considered for relative filenames)
+ * @param   {boolean}  [providedOptions.ignore]         False disables use of .eslintignore.
+ * @param   {string}   [providedOptions.ignorePath]     The ignore file to use instead of .eslintignore.
+ * @param   {string}   [providedOptions.ignorePattern]  A pattern of files to ignore.
+ * @param   {string}   [providedOptions.globInputPaths] False disables glob resolution.
  * @returns {string[]} Resolved absolute filenames.
  */
 function listFilesToProcess(globPatterns, providedOptions) {
     const options = providedOptions || { ignore: true };
+
+    if (options.globInputPaths === false) {
+        return globPatterns.map(filePath => ({ filename: filePath, ignored: false }));
```

Comment:
My assumption was that when `globInputPaths` is set to `false` the API callers would a properly pre-processed list of files, which mean expanding the globs, filtering out the ignore and computing the canonical path.

I guess it's ok to only skip globbing when `globInputPaths` is `false` and still verify the ignore files file as this should have a minimal performance impact.

However I wonder if we should also skip the call to `fs.realpathSync` [there](https://github.com/pvdlg/eslint/blob/cecfa4cf3fe8c9f4c2c74d1cb8a7956cd635dd0a/lib/util/glob-util.js#L165).
I would assume `fs.realpathSync` access the filesystem and has a performance cost.

I would prefer to skip it. And it seems reasonable to request/expect users to pass canonical file path when using `globInputPaths: false`. What do you think?

Reply:
Hmm, I understand the performance consideration, but skipping the `realpath` call seems unappealing to me because it could cause confusing behavior if an API user fails to read the documentation and passes a symlink instead. I think the result would be that the file would get read as usual, but the value of `context.getFilename()` in rules would be the path to the symlink rather than the realpath, and this could break plugins like `eslint-plugin-import` that rely on checking the current path to the file. As a result, the integration might appear to be working correctly but would actually be incorrect in confusing ways.

My impression is that there are other places where ESLint is not as efficient as it could be about filesystem accesses -- it seems like it would be better to improve those places rather than than adding a requirement on API users.

Reply:
@not-an-aardvark I made changes similar to what I was proposing in my last comment, including recent modifications merged into master.
Let me know if you'd like to make other changes.